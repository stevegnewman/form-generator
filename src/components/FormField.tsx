import React from 'react';
import { FormField as FormFieldType } from '../types/form';

interface FormFieldProps {
  field: FormFieldType;
  value: string;
  onChange: (name: string, value: string) => void;
}

export const FormField: React.FC<FormFieldProps> = ({ field, value, onChange }) => {
  if (field.type === 'dropdown') {
    return (
      <div className="mb-4">
        <label 
          htmlFor={field.name}
          className="block text-bodyText text-sm font-bold mb-2"
        >
          {field.label}
        </label>
        <select
          id={field.name}
          className="border rounded w-full py-2 px-3 text-bodyText
                     focus:ring-2 focus:ring-primary focus:border-primary
                     outline-none transition-all duration-200 cursor-pointer"
          value={value}
          onChange={(e) => onChange(field.name, e.target.value)}
        >
          <option value="">Select {field.label}</option>
          {field.options?.map((option) => (
            <option key={option} value={option}>
              {option}
            </option>
          ))}
        </select>
      </div>
    );
  }

  return (
    <div className="mb-4">
      <label 
        htmlFor={field.name}
        className="block text-bodyText text-sm font-bold mb-2">
        {field.label}
      </label>
      <input
        type="text"
        id={field.name}
        className="appearance-none border rounded w-full py-2 px-3 text-bodyText
                   focus:ring-2 focus:ring-primary focus:border-primary
                   outline-none transition-all duration-200"
        value={value}
        onChange={(e) => onChange(field.name, e.target.value)}
      />
    </div>
  );
}; 