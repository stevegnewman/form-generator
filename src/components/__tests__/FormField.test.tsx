import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { FormField } from '../FormField';

describe('FormField', () => {
  const mockOnChange = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders text input correctly', () => {
    const textField = {
      name: 'first_name',
      label: 'First Name',
      type: 'text' as const
    };

    render(<FormField field={textField} value="" onChange={mockOnChange} />);

    expect(screen.getByLabelText('First Name')).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('renders dropdown correctly', () => {
    const dropdownField = {
      name: 'country',
      label: 'Country',
      type: 'dropdown' as const,
      options: ['Canada', 'USA']
    };

    render(<FormField field={dropdownField} value="" onChange={mockOnChange} />);

    expect(screen.getByLabelText('Country')).toBeInTheDocument();
    expect(screen.getByRole('combobox')).toBeInTheDocument();
    expect(screen.getAllByRole('option')).toHaveLength(3); // Including default option
  });

  it('calls onChange when text input value changes', () => {
    const textField = {
      name: 'first_name',
      label: 'First Name',
      type: 'text' as const
    };

    render(<FormField field={textField} value="" onChange={mockOnChange} />);

    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'John' } });
    expect(mockOnChange).toHaveBeenCalledWith('first_name', 'John');
  });

  it('calls onChange when dropdown value changes', () => {
    const dropdownField = {
      name: 'country',
      label: 'Country',
      type: 'dropdown' as const,
      options: ['Canada', 'USA']
    };

    render(<FormField field={dropdownField} value="" onChange={mockOnChange} />);

    fireEvent.change(screen.getByRole('combobox'), { target: { value: 'Canada' } });
    expect(mockOnChange).toHaveBeenCalledWith('country', 'Canada');
  });
}); 