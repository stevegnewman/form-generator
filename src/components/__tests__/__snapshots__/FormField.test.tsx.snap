// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField matches snapshot for dropdown 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="block text-bodyText text-sm font-bold mb-2"
      for="test_dropdown"
    >
      Test Dropdown
    </label>
    <select
      class="border rounded w-full py-2 px-3 text-bodyText focus:ring-2 focus:ring-primary focus:border-primary outline-none transition-all duration-200 cursor-pointer"
      id="test_dropdown"
    >
      <option
        value=""
      >
        Select 
        Test Dropdown
      </option>
      <option
        value="Option 1"
      >
        Option 1
      </option>
      <option
        value="Option 2"
      >
        Option 2
      </option>
    </select>
  </div>
</div>
`;

exports[`FormField matches snapshot for text input 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="block text-bodyText text-sm font-bold mb-2"
      for="test_field"
    >
      Test Field
    </label>
    <input
      class="appearance-none border rounded w-full py-2 px-3 text-bodyText focus:ring-2 focus:ring-primary focus:border-primary outline-none transition-all duration-200"
      id="test_field"
      type="text"
      value=""
    />
  </div>
</div>
`;
